### ISSUE TITLE
Nullable column causes pipeline to fail in certain cases

### ISSUE THREAD

**Issue by steinitzu** [dlt-hub/dlt#126]
Hi hi

I'm just starting my contributions to dlt and ran into my first issue.

It happens when the column is `null` in first >=5000 rows (5000 matches buffer writer limit),
then a following row which has a non-null value for the column causes the pipeline to fail.

This pipeline reproduces the issue:

```python3
import dlt
import random

@dlt.source
def dummy():
    return [table_rows()]


@dlt.resource(write_disposition='append')
def table_rows():
    for _ in range(5000):
        yield dict(display_name=None, random_number=random.random())
    yield dict(display_name='Dummy name', random_number=random.random())


def main() -> None:
    p = dlt.pipeline(
        pipeline_name='dummy', destination='postgres', dataset_name='dummy_data', full_refresh=False
    )

    info = p.run(dummy())
    print(info)


if __name__ == '__main__':
    main()
```

Traceback:

```
Traceback (most recent call last):
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/runners/pool_runner.py", line 79, in run_pool
    run_metrics = run_f.run(cast(TPool, pool))
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/normalize/normalize.py", line 257, in run
    self.spool_schema_files(schema_name, list(files_in_schema))
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/normalize/normalize.py", line 235, in spool_schema_files
    self.spool_files(schema_name, load_id, map_parallel_f, files)
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/normalize/normalize.py", line 195, in spool_files
    total_items, schema_updates, chunk_files = map_f(schema, load_id, files)
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/normalize/normalize.py", line 171, in map_single
    processed_chunk = Normalize.w_normalize_files(
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/normalize/normalize.py", line 120, in w_normalize_files
    load_storage.close_writers(load_id)
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/storages/data_item_storage.py", line 33, in close_writers
    writer.close_writer()
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/data_writers/buffered.py", line 77, in close_writer
    self._flush_and_close_file()
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/data_writers/buffered.py", line 105, in _flush_and_close_file
    self._flush_items()
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/data_writers/buffered.py", line 100, in _flush_items
    self._writer.write_data(self._buffered_items)
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/data_writers/writers.py", line 151, in write_data
    write_row(rows[-1])
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/data_writers/writers.py", line 136, in write_row
    raise
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/data_writers/writers.py", line 136, in write_row
    raise
  File "/usr/lib/python3.10/bdb.py", line 90, in trace_dispatch
    return self.dispatch_line(frame)
  File "/usr/lib/python3.10/bdb.py", line 115, in dispatch_line
    if self.quitting: raise BdbQuit
bdb.BdbQuit
Traceback (most recent call last):
  File "/home/steinthor/projects/dlt-adrian/dlt/dlt/common/data_writers/writers.py", line 133, in write_row
    output[self._headers_lookup[n]] = self._caps.escape_literal(v)
KeyError: 'display_name'
```
Not sure what a solution is yet.  
Maybe a new file needs to be started when there are schema changes or something along those lines?

**Comment by rudolfix** [comment]
thanks for contributing! there's indeed a bug and your fix is right. the `columns` dict with the schema is shared by the buffered writer and the schema inference so the changes cannot be detected as this is the same object :) 
there's no test for that case I'll work a little bit on #127 to add tests and then merge it

if you are blocked by this you can run your script like `BUFFER_MAX_ITEMS=1000000 python pipeline.py` or add this config value to `config.toml` in `.dlt` folder.
```toml
buffer_max_items=1000000
```
until we merge and release the bugfix

**Comment by rudolfix** [comment]
fixed by #129
